doc/design/high-level.md

```diff
@@ -0,0 +1,214 @@
+the build. Xcode Build Settings, target dependencies, etc. are all set up to
+create a normal Xcode experience.
+
+To ensure that the resulting build is as similar to a Bazel build as
+possible, some things are done differently than a vanilla Xcode project. In
+particular, `BUILT_PRODUCTS_DIR` is set to a nested folder, mirroring the layout
+of `bazel-out/`, and various search paths are adjusted to account for this.
+
+There are also aspects of a Bazel build that can't be neatly translated into an
+Xcode concept (though updating rules to supply [rules_xcodeproj
+providers](#providers) can help). One that will come up in nearly every project
+is code generation. In these situations the project takes on a hybrid approach,
+invoking `bazel build` for a portion of the build. The degree to which `bazel
+build` needs to handle the build depends on the the rules involved.
+
+Of note, the project can be customized to force more of it to be hybrid, through
+the use of [focused project rules](#additional-rules).
+
+### Build with Bazel
+
+In the "Build with Bazel" mode, the generated project will invoke `bazel build`
+to perform the actual build, as a root or detached dependency, and then stub
+out the build actions that Xcode tries to perform. A version of this method is
+detailed [here](https://github.com/kastiglione/bazel-xcode-demo-swift-driver/tree/master/bazel#disabling-xcodes-build)
+and [here](https://www.youtube.com/watch?t=1301&v=NAPeWoimGx8).
+
+In addition to the stubbing, targets will have their serialized diagnostics
+replayed, resulting in fully functional warnings and fix-its, that stick around
+between builds, but also clear when they are supposed to.
+
+This will be the default build mode by the 1.0 release.
+
+### Build with Bazel via Proxy
+
+rules_xcodeproj will support one more build mode called "Build with Bazel via
+Proxy". In this mode the generated project will rely on Xcode using an
+[XCBBuildServiceProxy](https://github.com/target/XCBBuildServiceProxy). This
+takes the Xcode build system entirely out of the equation, allowing Bazel to
+fully control the build.
+
+Here are some benefits that the proxy provides over "Build with Bazel":
+
+- No additional "Bazel Dependencies" target in the build graph
+- Removal of duplicate warnings/errors
+- More stable Indexing
+- User created schemes (without defining bazel rules to create them) work as
+  expected
+- Fully functional progress bar (though there is hope that we can get it to
+  partially work with "Build with Bazel")
+- Less overhead
+
+With all of these benefits, you may be wondering why "Build with Bazel" will be
+the default build mode instead of this one. There are two main reasons:
+
+- The proxy relies on a private API that Xcode uses to communicate with
+  XCBBuildService. This can have, and has had, breaking changes between Xcode
+  versions.
+- To have Xcode use the proxy, you either need to launch it with an environment
```

Comment:
I'd add what the consequence of this is, which as I understand it, is that it would force all projects opened with that instance of Xcode to use XCBBuildServiceProxy and build with bazel, which could be much more inconvenient than launching Xcode with a specific command.

Reply:
The proxy is written to only build with bazel for projects generated with rules_xcodeproj. Currently though proxies are written to only work with a single version of Xcode, so setting the environment variable is probably too large of a change. I'm sure we could make the proxy work with a range of Xcode versions though.

I'll add this to the docs :+1:.